### DISCUSSION TITLE
Why is android:extractNativeLibs removed while merging splits?

### DISCUSSION THREAD

**Discussion by 84436** [REAndroid/APKEditor#150] [Q&A] [answered]
Hi everyone. I notice that when I'm merging splits, there's a line in the log that says the `extractNativeLibs` is removed from the APK's manifest. Why does that attribute have to be removed?

Note that I'm just wondering the behavior of APKEditor split merger; I have no issues with the merged APK whatsoever. It's just that I have read [Android's docs for the `<application>` tag in `AndroidManifest.xml`](https://developer.android.com/guide/topics/manifest/application-element#extractNativeLibs), and if I understand correctly, setting `extractNativeLibs` to `false` cause the `*.so` files in `lib\$ABI` folder of the APK to be read directly by the Android runtime instead of them having to be extracted first by the Package Installer, thus saving disk space.

For reference, I'm using APKEditor to merge the base APK and 5 splits of my selection (`arm64_v8a`, `en`, `vi`, `xhdpi`, `xxhdpi`) of the YouTube app, with the intention of applying ReVanced patches against the merged APK later. Here's a shot of my terminal:

```
( 0:20:00.11) D:\ReVanced\youtube-19.46.42
> apkeditor m -i splits -o original.apk -validate-modules 
00.000 I: [MERGE] Using: APKEditor version 1.4.1, ARSCLib version 1.3.5
      -validate-modules = true
                     -i = splits
                     -o = original.apk
 ____________________________________
00.013 I: [MERGE] Searching apk files ...
00.014 I: [MERGE] Found apk files: 6
00.276 I: [MERGE] Found modules: 6
00.358 I: [MERGE] Merging: base
00.362 I: [MERGE] Added [base] classes.dex -> classes.dex  
00.362 I: [MERGE] Added [base] classes2.dex -> classes2.dex
00.362 I: [MERGE] Added [base] classes3.dex -> classes3.dex
00.362 I: [MERGE] Added [base] classes4.dex -> classes4.dex
00.363 I: [MERGE] Added [base] classes5.dex -> classes5.dex
00.363 I: [MERGE] Added [base] classes6.dex -> classes6.dex
00.363 I: [MERGE] Added [base] classes7.dex -> classes7.dex
01.187 I: [MERGE] Merging: split_config.vi
01.413 I: [MERGE] Merging: split_config.xhdpi
01.593 I: [MERGE] Merging: split_config.xxhdpi
01.765 I: [MERGE] Merging: split_config.arm64_v8a
01.773 I: [MERGE] Merging: split_config.en
02.006 I: [MERGE] Sanitizing manifest ...
02.010 I: [MERGE] Removed-attribute : @0x0101064e
02.010 I: [MERGE] Removed-attribute : @0x0101064f
02.013 I: [MERGE] Removed-attribute : extractNativeLibs
02.013 I: [MERGE] Removed-element : <meta-data> name="com.android.vending.splits.required"
02.013 I: [MERGE] Removed-element : <meta-data> name="com.android.stamp.source"
02.013 I: [MERGE] Removed-element : <meta-data> name="com.android.stamp.type"
02.019 I: [MERGE] Removed-table-entry : res/xml/splits0.xml
02.019 I: [MERGE] Removed-element : <meta-data> name="com.android.vending.splits"
02.019 I: [MERGE] Removed-element : <meta-data> name="com.android.vending.derived.apk.id" 
02.180 I: [MERGE] Writing apk ...
02.203 I: [MERGE] Buffering compress changed files ...
07.800 I: [MERGE] Writing files: 8360
07.982 I: [MERGE] Writing signature block ...
08.017 I: [MERGE] Saved to: original.apk

( 0:20:00.11) D:\ReVanced\youtube-19.46.42
>
```

And here are the version numbers for posterity's sake:
- APKEditor, version 1.4.1
- Java JDK provided by Eclipse Temurin, build 21.0.3+9-LTS
- YouTube APK bundle, version 19.46.42
- Windows 10, build 19044.5131

**Comment by REAndroid** [comment, marked as answer]
@84436 

Oh sorry I don't know how I didn't see this for long, i blame github not having some kind of notifications.

Yes you are correct about `extractNativeLibs` attribute function
| extractNativeLibs        | Behavior on android OS           | Storage/compression on apk file | Pros/Cons |
| ------------- |:-------------:| :-----:|:-----:|
|false  |Reads and executes libs directly from apk| Stored uncompressed with 4096 alignment | Bigger apk file size, Smaller phone/device storage usage|
| true or removed |Extracts lib files to storage| Stored compressed | Smaller apk file size, Bigger phone/device storage usage|

We (APKEditor) choose `false` by default but i feel we have to let the user decide which option to choose just like AGP offering with `useLegacyPackaging` flag

**Comment by REAndroid** [comment]
Now `extractNativeLibs` is fully implemented as on [V1.4.3](https://github.com/REAndroid/APKEditor/releases/tag/V1.4.3)
